lib/SILOptimizer/Transforms/PhiArgumentOptimizations.cpp

```diff
@@ -93,6 +93,17 @@ void RedundantPhiEliminationPass::run() {
   }
 }
 
+static bool hasOnlyNoneOwnershipIncomingValues(SILArgument *phi) {
```

Comment:
I'm not sure if this check also works in case the none -> owned "conversion" happens not in the _immediate_ predecessor. For example, when I modify your test case like this, it's done in the predecessor of bb1 and not in the predecessor of bb3:
```
bb0:
   %0 = enum $FakeOptional<Klass>, #FakeOptional.none!enumelt 
   cond_br undef, bb1(%0 : $FakeOptional<Klass>, %0 : $FakeOptional<Klass>), bb2

 bb1(%x : @owned $FakeOptional<Klass>, %y : @owned $FakeOptional<Klass>):
   br bb3(%x : $FakeOptional<Klass>, %y : $FakeOptional<Klass>)

 bb2:
   br bb3(%0 : $FakeOptional<Klass>, %0 : $FakeOptional<Klass>)

bb3(%1 : @owned $FakeOptional<Klass>, %2 : @owned $FakeOptional<Klass>):
   ...
```
Maybe you can verify this condition in `valuesAreEqual`?

Also, if you are using this static function only in asserts, please put a `#ifndef NDEBUG` around it. Otherwise there will be a warning in release builds.

Reply:
Right, thanks. Also there's some additional cases where I have to insert lifetime ends explicitly. I'll work on this and update the PR.

Reply:
I think its fine to allow none -> owned conversion via an intermediate @owned phi. As long as we correctly insert a compensating destroy_value. I changed `hasOnlyNoneOwnershipIncomingValues` to reflect this. But not disabling this case in `valuesAreEqual `